### PR TITLE
HOTT-1628: Move to using materialized path to generate the gn csv hierarchy

### PR DESCRIPTION
**Pull request by willfish**
### Jira link

https://transformuk.atlassian.net/browse/HOTT-1628

### What?

I have added/removed/altered:

- [x] Added hierarchy using the materialized path in the admin goods nomenclature csv endpoint
- [x] Moved to asynchronous persistence model for generation of daily goods nomenclature file

### Why?

I am doing this because:

- This is required to make it easier and faster for us to retrieve a today's goods nomenclature file

### CODE REVIEW

File: app/serializers/api/admin/csv/goods_nomenclature_serializer.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        column :goods_nomenclature_sid, column_name: 'SID'
+        column :goods_nomenclature_item_id, column_name: 'Commodity code'
+        column :producline_suffix, column_name: 'Product line suffix'
+        column :description, column_name: 'Description'
+        column :validity_start_date, column_name: 'Start date'
+        column :validity_end_date, column_name: 'End date'
+        column :number_indents, column_name: 'Indentation'
+        column :end_line, column_name: 'End line', &:declarable?
+
+        # Uses materialized path to determine declarability of goods nomenclature to avoid slow csv generation
```
[the remark this answers is not in the document]
Comment (willfish) [reply app/serializers/api/admin/csv/goods_nomenclature_serializer.rb:16]: One for the laterbase

File: app/lib/tariff_synchronizer/file_service.rb [hunk 2]
```diff
@@ -1,23 +1,17 @@
 module TariffSynchronizer
   class FileService
     class << self
-      # Uploads local file to S3 bucket
-      def upload_file(src_path, dst_path)
```
Comment (willfish) [inline app/lib/tariff_synchronizer/file_service.rb:5]: Delete and upload file aren't used

File: app/serializers/concerns/api/shared/csv_serializer.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        CSV.generate do |csv|
+          serializable_array.each do |row|
+            csv << row
+          end
+        end
+      end
+
+      delegate :column_options, to: :class
+
+        private
```
Comment (alexdesi) [inline app/serializers/concerns/api/shared/csv_serializer.rb:35]: strange indentation here

File: app/serializers/concerns/api/shared/csv_serializer.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        column_options.pluck(:column_name)
+      end
+
+      module ClassMethods
+        def columns(*column_options)
+          column_options.each(&method(:column))
+        end
+
+        def column(column, options = {}, &block)
+          (self.column_options ||= []) << options.reverse_merge(column:, column_name: column, value_block: block)
```
Comment (alexdesi) [inline app/serializers/concerns/api/shared/csv_serializer.rb:47]: maybe in two rows this is shorter and more readable
```
self.column_options ||= []
self.column_options << ...
```

File: app/workers/generate_goods_nomenclatures_report_worker.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
+class GenerateGoodsNomenclaturesReportWorker
```
Comment (alexdesi) [inline app/workers/generate_goods_nomenclatures_report_worker.rb:1]: report is a bit generic
what about
`GenerateGoodsNomenclaturesCsvWorker` ?
it's even easier to search

Comment (willfish) [reply app/workers/generate_goods_nomenclatures_report_worker.rb]: Done

**Pull request merged**